dnf/util.py

```diff
@@ -25,22 +25,186 @@
 from .pycomp import PY3, basestring
 from functools import reduce
 
+
+##from dnfpluginscore import _, logger
+from dnf.i18n import _
+import dnf
+import iniparse
+
 import dnf.const
 import dnf.pycomp
 import itertools
 import librepo
+import logging
 import os
 import pwd
 import shutil
 import subprocess
 import tempfile
 import time
 
-"""DNF Utilities.
+logger = logging.getLogger('dnf')
+
+"""DNF Utilities."""
+
+
+def get_reposdir(plugin):
+    """
+    # :api
+    Returns the value of reposdir
+    """
+    myrepodir = None
+    # put repo file into first reposdir which exists or create it
+    for rdir in plugin.base.conf.reposdir:
+        if os.path.exists(rdir):
+            myrepodir = rdir
+
+    if not myrepodir:
+        myrepodir = plugin.base.conf.reposdir[0]
+        ensure_dir(myrepodir)
+    return myrepodir
+
+
+def _non_repo_handle(conf):
+    handle = librepo.Handle()
+    handle.useragent = dnf.const.USER_AGENT
+    # see dnf.repo.Repo._handle_new_remote() how to pass
+    handle.maxspeed = conf.throttle if type(conf.throttle) is int \
+        else int(conf.bandwidth * conf.throttle)
+    handle.proxy = conf.proxy
+    handle.proxyuserpwd = dnf.repo._user_pass_str(conf.proxy_username,
+                                                  conf.proxy_password)
+    handle.sslverifypeer = handle.sslverifyhost = conf.sslverify
+    return handle
+
+
+def _build_default_handle():
+    handle = librepo.Handle()
+    handle.useragent = dnf.const.USER_AGENT
+    return handle
+
+
+def urlopen(url, conf=None, repo=None, mode='w+b', **kwargs):
+    """
+    # :api
+    Open the specified absolute url, return a file object
+    which respects proxy setting even for non-repo downloads
+    """
+    if PY3 and 'b' not in mode:
+        kwargs.setdefault('encoding', 'utf-8')
+    fo = tempfile.NamedTemporaryFile(mode, **kwargs)
+    if repo:
+        handle = repo.get_handle()
+    elif conf:
+        handle = _non_repo_handle(conf)
+    else:
+        handle = _build_default_handle()
+    try:
+        librepo.download_url(url, fo.fileno(), handle)
+    except librepo.LibrepoException as e:
+        raise IOError(e.args[1])
+    fo.seek(0)
+    return fo
+
+
+def write_raw_configfile(filename, section_id, substitutions, modify):
```

Comment:
dnf.conf.Conf =  dnf.yum.config.YumConf
Somehow I am not sure about this destination in YumConf class.